File: day12-java/Day.java

import java.util.ArrayList;

public class Day {

        private static String[] sampleInput = new String[] {
                "cpy 41 a",
                "inc a",
                "inc a",
                "dec a",
                "jnz a 2",
                "dec a"
        };

        private static String[] input = new String[] {
                "cpy 1 a",
                "cpy 1 b",
                "cpy 26 d",
                "jnz c 2",
                "jnz 1 5",
                "cpy 7 c",
                "inc d",
                "dec c",
                "jnz c -2",
                "cpy a c",
                "inc a",
                "dec b",
                "jnz b -2",
                "cpy c b",
                "dec d",
                "jnz d -6",
                "cpy 19 c",
                "cpy 14 d",
                "inc a",
                "dec d",
                "jnz d -2",
                "dec c",
                "jnz c -5"
        };

        public static void main(String[] args) {

                Memory memory = new Memory();
                System.out.println(memory);

                for (int idx = 0; idx < input.length; idx++) {

                        String rawInstruction = input[idx];

                        String[] instruction = rawInstruction.split(" ");
                        if (instruction[0].equals("cpy")) {
                                try {
                                        memory.set(Integer.parseInt(instruction[1]), instruction[2].charAt(0));
                                } catch(NumberFormatException e) {
                                        memory.set(instruction[1].charAt(0), instruction[2].charAt(0));
                                }
                        } else if (instruction[0].equals("inc")) {
                                memory.increase(instruction[1].charAt(0));
                        } else if (instruction[0].equals("dec")) {
                                memory.decrease(instruction[1].charAt(0));
                        } else if (instruction[0].equals("jnz")) {
                                int jumXValue = 0;
                                try {
                                        jumXValue = Integer.parseInt(instruction[1]);
                                } catch(NumberFormatException e) {
                                        jumXValue = memory.get(instruction[1].charAt(0));
                                }

                                if (jumXValue == 0) {
                                        // We can ignore this jump
                                        continue;
                                }

                                idx += Integer.parseInt(instruction[2]);
                                idx--;
                                continue;

                        } else {
                                throw new IllegalArgumentException("Not a valid instruction :: " + rawInstruction);
                        }

                }

                System.out.println(memory);

        }

}

File: day23-php/day.php
<?php

        include_once 'memory.php';

        $memory = new Memory();

        $input_test = array(
                "cpy 2 a",
                "tgl a",
                "tgl a",
                "tgl a", //inc a // a=
                "cpy 1 a", //jnz 1 a
                "dec a", //a=0
                "dec a" // a=-1
        );

        $input = array(
                "cpy a b",
                "dec b",
                "cpy a d",
                "cpy 0 a",
                "cpy b c",
                "inc a",
                "dec c",
                "jnz c -2",
                "dec d",
                "jnz d -5",
                "dec b",
                "cpy b c",
                "cpy c d",
                "dec d",
                "inc c",
                "jnz d -2",
                "tgl c",
                "cpy -16 c",
                "jnz 1 c",
                "cpy 89 c",
                "jnz 90 d",
                "inc a",
                "inc d",
                "jnz d -2",
                "inc c",
                "jnz c -5"
        );

        echo $memory;

        for ($idx=0;$idx<count($input);$idx++) {

                //echo("\n\nexecute ${idx} ${input[$idx]}");
                //echo("\n".implode($input, ","));
                //echo("\n${memory}");

                $instruction = explode(" ", $input[$idx]);
                if (strcmp($instruction[0], "cpy") == 0) {

                        if (is_numeric($instruction[1])) {
                                $memory->set($instruction[1], $instruction[2]);
                        } else {
                                $memory->setValueFromRegister($instruction[1], $instruction[2]);
                        }
                } else if (strcmp($instruction[0], "inc") == 0) {
                        $memory->increase($instruction[1]);
                } else if (strcmp($instruction[0], "dec") == 0) {
                        $memory->decrease($instruction[1]);
                } else if (strcmp($instruction[0], "jnz") == 0) {
                        $jumXValue = 0;
                        if (is_numeric($instruction[1])) {
                                $jumXValue = $instruction[1];
                        } else {
                                $jumXValue = $memory->get($instruction[1]);
                        }

                        if ($jumXValue == 0) {
                                // We can ignore this jump
                                continue;
                        }

                        $jumYValue = 0;
                        if (is_numeric($instruction[2])) {
                                $jumYValue = $instruction[2];
                        } else {
                                $jumYValue = $memory->get($instruction[2]);
                        }

                        $idx += $jumYValue;
                        $idx--;
                } else if (strcmp($instruction[0], "tgl") == 0) {

                        // The instruction that will be written
                        // to the instruction that is to be toggled
                        $newInstruction = "";

                        $toggleValue = 0;
                        if (is_numeric($instruction[1])) {
                                $toggleValue = $instruction[1];
                        } else {
                                $toggleValue = $memory->get($instruction[1]);
                        }

                        if(($idx + $toggleValue) > (count($input)-1)) {
                                continue;
                        }

                        $instructionToToggle = explode(" ", $input[$idx + $toggleValue]);

                        //one-argument instructions
                        if (count($instructionToToggle) == 2) {
                                if (strcmp($instructionToToggle[0], "inc") == 0) {
                                        $instructionToToggle[0] = "dec";
                                } else {
                                        $instructionToToggle[0] = "inc";
                                }
                        }

                        // two-argument instructions
                        else if (count($instructionToToggle) == 3) {
                                if (strcmp($instructionToToggle[0], "jnz") == 0) {
                                        $instructionToToggle[0] = "cpy";
                                } else {
                                        $instructionToToggle[0] = "jnz";
                                }
                        }
                        $input[$idx + $toggleValue] = implode($instructionToToggle, " ");

                } else {
                        throw new Exception("Not a valid instruction :: '".implode($instruction, " ")."' at index ".$idx);
                }
        }

        echo("\n${memory}");

?>
